feature/amazon/src/main/java/com/revenuecat/purchases/amazon/AmazonBilling.kt

```diff
@@ -140,31 +187,187 @@ class AmazonBilling constructor(
+    // compile as long as all of the functions are implemented, otherwise it doesn't know which delegated
+    // implementation to take
+    override fun onUserDataResponse(response: UserDataResponse) {
+        userDataHandler.onUserDataResponse(response)
+    }
+
+    override fun onProductDataResponse(response: ProductDataResponse) {
+        productDataHandler.onProductDataResponse(response)
+    }
+
+    override fun onPurchaseResponse(response: PurchaseResponse) {
+        purchaseHandler.onPurchaseResponse(response)
+    }
+
+    override fun onPurchaseUpdatesResponse(response: PurchaseUpdatesResponse) {
+        purchaseUpdatesHandler.onPurchaseUpdatesResponse(response)
+    }
+
+    private fun logErrorsIfAny(errors: Map<String, PurchasesError>) {
+        if (errors.isNotEmpty()) {
+            val receiptsWithErrors = errors.keys.joinToString("\n")
+            log(
+                LogIntent.AMAZON_ERROR,
+                AmazonStrings.ERROR_FETCHING_RECEIPTS.format(receiptsWithErrors)
+            )
+        }
+    }
+
+    private fun getMissingSkusForReceipts(
+        appUserID: String,
+        amazonUserID: String,
+        receipts: List<Receipt>,
+        onCompletion: (tokensToSkusMap: Map<String, String>, errors: Map<String, PurchasesError>) -> Unit
+    ) {
+        val currentlyCachedTokensToSkus: Map<String, String> = cache.getReceiptSkus()
+
+        val successMap: MutableMap<String, String> = currentlyCachedTokensToSkus.toMutableMap()
+        val errorMap: MutableMap<String, PurchasesError> = mutableMapOf()
+
+        val nonSubscriptionReceiptsToSku =
+            receipts.filterNot { it.productType == ProductType.SUBSCRIPTION }
+                .map { it.receiptId to it.sku }
+
+        successMap.putAll(nonSubscriptionReceiptsToSku)
+
+        val subscriptionReceiptsToFetchTermSku: List<Receipt> =
+            receipts
+                .filter { it.productType == ProductType.SUBSCRIPTION }
+                .filterNot { currentlyCachedTokensToSkus.containsKey(it.receiptId) }
+
+        if (subscriptionReceiptsToFetchTermSku.isEmpty()) {
+            onCompletion(successMap, errorMap)
+            return
+        }
+
+        var receiptsLeft = subscriptionReceiptsToFetchTermSku.count()
+        subscriptionReceiptsToFetchTermSku.forEach { receipt ->
+            amazonBackend.getAmazonReceiptData(
+                receipt.receiptId,
+                amazonUserID,
+                onSuccess = { response ->
+                    log(LogIntent.DEBUG, AmazonStrings.RECEIPT_DATA_RECEIVED.format(response.toString()))
+
+                    successMap[receipt.receiptId] = response["termSku"] as String
+
+                    receiptsLeft--
+                    if (receiptsLeft == 0) {
+                        cache.setReceiptSkus(successMap)
+                        onCompletion(successMap, errorMap)
+                    }
+                }
+            ) { error ->
```

Comment:
I found the indentation here a bit confusing because we're using a named block for success but an implicit block for error, maybe doing 
```kotlin
amazonBackend.getAmazonReceiptData(
    receipt.receiptId,
    amazonUserID,
    onSuccess = {
    ...
    }, onError = { 
    ...
    })
```
would make things clearer